text/0000-amcl-replacement.md

```diff
@@ -0,0 +1,41 @@
+# Cartographer AMCL replacement
+
+## Summary
+[summary]: #summary
+
+Make cartographer in Pure localization mode works as a replacement for AMCL and be one of the ROS localization method.
+
+## Motivation
+[motivation]: #motivation
+
+Cartographer is really powerful in term of localization and a pure localization mode was developed (it works already great). In order to be adopted by the ROS community it should be simple to run and works similarly  in its implementation as AMCL
+
+
+## Approach
+[approach]: #approach
+
+* Use parameters to initialize the starting pose of the robot, similarly as ` ~initial_pose_x`, ` ~initial_pose_y`, `~initial_pose_a` in AMCL: http://wiki.ros.org/amcl
+
+
+* Support the `/initialpose` topic to relocate the robot. 
+This topic is for instance used out of the box by rviz `2D pose estimate` button (or keyboard shortcut: p). 
+A quick and dirty implementation was done here: https://github.com/googlecartographer/cartographer_ros/pull/637
+
+
+* Publish the static map against which localization is done on the `/map` topic for the navigation stack to be able to be run on top of it (global planner for instance) and to have a static reference to record waypoint in.
+Currently with the occupancy_grid node, the map which is published is a superposition of the static map + the current submap. 
+A workaround is to not run the occupancy_grid node and in parallel run a `map_server` node which publish a *.pgm/*.png/ + *.yaml map that had to be generated from the same *.pbstream file which used by the pure localization node. 
+It would be nice to have everything work in one step.
+
+
+* Provide a way to easily edit a map afterwards (if this is possible). 
+AMCL maps being bitmaps, there are easily cleaned manually, in GIMP for instance. 
+This is very useful when maping a site which has moveable furnitures (chairs, people standing, christmas tree, etc...), we want to be able to remove these non fix objects from the localization map.
+
+
+* Make the origin of the map changeable. 
```

Comment:
Not sure why this is in here, there is no motivation for this provided. What does the origin mean even?

Reply:
Sorry for being unclear. Currently with cartographer in localization mode, the pose of the robot is defined in the map frame relative to a point with the (0 0) coordinate. This point being located on the map to where the robot started the maping procedure (ie where the robot was when the bag used to generate the .pbstream started). This is not very conveniant as every other point of interest have to be defined relative to this origin, and this origin will change if a new map is built. A workaround is to publish a tf from the origin of the cartographer map to a choosen physical point and then to record any point of interest relative to this physical point. But I think this could be simplified.

Reply:
@doisyg Can you make a concrete proposal what you want to modify in the code?

Reply:
@sebastianklose This could be another feature for the pbstream tool we discussed.

Reply:
How should this work? 

1. Start mapping / localizing at an initial pose that's relative to a world-fixed parent frame - for me this is not a workaround as you say, but the only way to do it right. Your motivation sounds like you have such a global reference frame anyway, otherwise it wouldn't matter?

2. You can't simply change the trajectory to start at an arbitrary point without adding a proper parent frame afterwards - either you move the whole trajectory inside the map (semantically wrong) or keep it as it is and add an additional starting node (adds a jump that can be problematic for optimization).

Reply:
```
1.    Start mapping / localizing at an initial pose that's relative to a world-fixed parent frame - for me this is not a workaround as you say, but the only way to do it right. Your motivation sounds like you have such a global reference frame anyway, otherwise it wouldn't matter?
```
I am not sure I understand fully your comment. But we do not necessarily have a global reference frame. I know a couple of use cases (including ours) where a map is build first, then coordinates of point of interests are recorded/defined . Is that case, the ROS map is the global reference frame (or its origin). The advantage of the ROS map implementation is that by default, the origin of the map is the corner of the map, and that is it is very easy to change where this 0 is located with the companion yaml file. http://wiki.ros.org/map_server#YAML_format

```
2.    You can't simply change the trajectory afterwards to start at an arbitrary point without adding a proper parent frame - either you move the whole trajectory inside the map (semantically wrong) or keep it as it is and add an additional starting node (adds a jump that can be problematic for optimization). Or do I get it wrong?

```
As I said before, a ROS user could add a tf transform on top of cartographer to get a convenient 0 on its 'map_relocated' frame, but I feel that there should be an easier way.
For instance, I can imagine 3 parameters (x,y, theta) that the cartographer_ros node will use to offset the 0 of the map when publishing the robot pose in ROS.

Reply:
What I mean is that you always need an additional frame in which your SLAM trajectories are embedded.

> a ROS user could add a tf transform on top of cartographer to get a convenient 0 on its 'map_relocated' frame, but I feel that there should be an easier way.
For instance, I can imagine 3 parameters (x,y, theta) that the cartographer_ros node will use to offset the 0 of the map when publishing the robot pose in ROS

The only difference would be that the additional TF is provided by the `cartographer_node`, which is a kind of special thing to add to its options. Just use a `static_transform_publisher` to your launch file - done.